Spanner/tests/System/PgOperationsTest.php

```diff
@@ -0,0 +1,131 @@
+{
+    private static $row = [];
+
+    private static $id1;
+
+    public static function setUpBeforeClass(): void
+    {
+        parent::setUpBeforeClass();
+
+        self::$id1 = rand(1000, 9999);
+        self::$row = [
+            'id' => self::$id1,
+            'name' => uniqid(self::TESTING_PREFIX),
+            'birthday' => new Date(new \DateTime('2000-01-01'))
+        ];
+
+        self::$database->insert(self::TEST_TABLE_NAME, self::$row);
+    }
+
+    public function testInsertWithDbRole()
+    {
+        // Emulator does not support FGAC
+        $this->skipEmulatorTests();
+
+        foreach ($this->insertDbProvider() as $dbProvided) {
+            list($db, $values, $expected) = $dbProvided;
+            $error = null;
+
+            try {
+                $res = $db->insert(self::TEST_TABLE_NAME, $values);
+            } catch (ServiceException $e) {
+                $error = $e;
+            }
+
+            if ($expected === null) {
+                $this->assertEquals($error, $expected);
+            } else {
+                $this->assertInstanceOf(ServiceException::class, $error);
+                $this->assertEquals($error->getServiceException()->getStatus(), $expected);
+            }
+        }
+    }
+
+    public function testReadWithDbRole()
+    {
+        // Emulator does not support FGAC
+        $this->skipEmulatorTests();
+
+        foreach ($this->readDbProvider() as $dbProvided) {
+            list($db, $expected) = $dbProvided;
+            $error = null;
+            $keySet = self::$client->keySet([
+                'keys' => [self::$id1]
+            ]);
+            $columns = ['id', 'name', 'birthday'];
+
+            try {
+                $res = $db->read(self::TEST_TABLE_NAME, $keySet, $columns);
+                $row = $res->rows()->current();
+            } catch (ServiceException $e) {
+                $error = $e;
+            }
+
+            if ($expected === null) {
+                $this->assertEquals(self::$id1, $row['id']);
+            } else {
+                $this->assertInstanceOf(ServiceException::class, $error);
+                $this->assertEquals($error->getServiceException()->getStatus(), $expected);
+            }
+        }
+    }
+
+    public function insertDbProvider()
+    {
+        return [
+            [
+                self::$dbWithRestrictiveRole,
+                [
+                    'id' => rand(1, 346464),
+                    'name' => uniqid(self::TESTING_PREFIX),
+                    'birthday' => new Date(new \DateTime('2000-01-01'))
+                ],
+                'PERMISSION_DENIED'
+            ],
+            [
+                self::$dbWithSessionPoolRestrictiveRole,
+                [
+                    'id' => rand(1, 346464),
+                    'name' => uniqid(self::TESTING_PREFIX)
+                ],
+                null
+            ]
+        ];
+    }
+
+    public function readDbProvider()
```

Comment:
Same for `readDbProvider` it can be abstracted out in a trait or testhelper class.

Reply:
Great suggestion, Thank you, Done.